--- wpa_supplicant-2.7/wpa_supplicant/wpa_cli.c	2018-12-03 03:34:59.000000000 +0800
+++ wpa_supplicant-2.7_new/wpa_supplicant/wpa_cli.c	2021-03-01 15:15:22.741735350 +0800
@@ -53,6 +53,7 @@ static const char *pid_file = NULL;
 static const char *action_file = NULL;
 static int ping_interval = 5;
 static int interactive = 0;
+static int scan_autoquit = 0;
 static char *ifname_prefix = NULL;
 
 static DEFINE_DL_LIST(bsses); /* struct cli_txt_entry */
@@ -89,6 +90,7 @@ static void usage(void)
 	       "  -a = run in daemon mode executing the action file based on "
 	       "events from\n"
 	       "       wpa_supplicant\n"
+	       "  -A = stop action file after finish scan"
 	       "  -B = run a daemon in the background\n"
 	       "  default path: " CONFIG_CTRL_IFACE_DIR "\n"
 	       "  default interface: first interface found in socket path\n");
@@ -2394,6 +2396,7 @@ static char ** wpa_cli_complete_p2p_set(
 		"managed",
 		"listen_channel",
 		"ssid_postfix",
+		"passphrase",
 		"noa",
 		"ps",
 		"oppps",
@@ -3927,6 +3930,9 @@ static void wpa_cli_action_process(const
 		wpa_cli_exec(action_file, ctrl_ifname, pos);
 	} else if (str_starts(pos, MESH_PEER_DISCONNECTED)) {
 		wpa_cli_exec(action_file, ctrl_ifname, pos);
+	} else if (str_starts(pos, "P2P")) {
+		/* Added by Ambarella for hook all P2P commands. */
+		wpa_cli_exec(action_file, ifname, pos);
 	} else if (str_starts(pos, P2P_EVENT_GROUP_STARTED)) {
 		wpa_cli_exec(action_file, ifname, pos);
 	} else if (str_starts(pos, P2P_EVENT_GROUP_REMOVED)) {
@@ -3960,6 +3966,8 @@ static void wpa_cli_action_process(const
 	} else if (str_starts(pos, WPA_EVENT_TERMINATING)) {
 		printf("wpa_supplicant is terminating - stop monitoring\n");
 		wpa_cli_quit = 1;
+	//} else {
+		//syslog(0, "%s:%s", __func__, pos);
 	}
 }
 
@@ -4093,8 +4101,13 @@ static void wpa_cli_recv_pending(struct
 		size_t len = sizeof(buf) - 1;
 		if (wpa_ctrl_recv(ctrl, buf, &len) == 0) {
 			buf[len] = '\0';
-			if (action_monitor)
+			if (action_monitor) {
+				if (scan_autoquit && (NULL != strstr(buf, WPA_EVENT_SCAN_RESULTS))) {
+					//syslog(0, "%s %s", __func__, buf);
+					eloop_terminate();
+				}
 				wpa_cli_action_process(buf);
+			}
 			else {
 				cli_event(buf);
 				if (wpa_cli_show_event(buf)) {
@@ -4540,10 +4553,14 @@ int main(int argc, char *argv[])
 		return -1;
 
 	for (;;) {
-		c = getopt(argc, argv, "a:Bg:G:hi:p:P:s:v");
+		c = getopt(argc, argv, "A:a:Bg:G:hi:p:P:s:v");
 		if (c < 0)
 			break;
 		switch (c) {
+		case 'A':
+			scan_autoquit = 1;
+			action_file = optarg;
+			break;
 		case 'a':
 			action_file = optarg;
 			break;
@@ -4652,11 +4669,21 @@ int main(int argc, char *argv[])
 		if (daemonize && os_daemonize(pid_file) && eloop_sock_requeue())
 			return -1;
 
+		if (scan_autoquit) {
+			char *scan_arg[] = { "scan" };
+			ret = wpa_request(ctrl_conn, 1, &scan_arg[0]);
+		}
+
 		if (action_file)
 			wpa_cli_action(ctrl_conn);
 		else
 			ret = wpa_request(ctrl_conn, argc - optind,
 					  &argv[optind]);
+
+		if (scan_autoquit) {
+			char *scan_arg[] = { "scan_r" };
+			ret = wpa_request(ctrl_conn, 1, &scan_arg[0]);
+		}
 	}
 
 	os_free(ctrl_ifname);
