diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index 179cf43..05eb915 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -35,6 +35,7 @@
 #include "wnm_ap.h"
 #include "mbo_ap.h"
 #include "ndisc_snoop.h"
+#include <syslog.h>
 #include "sta_info.h"
 #include "vlan.h"
 #include "wps_hostapd.h"
@@ -1165,6 +1166,17 @@ void ap_sta_stop_sa_query(struct hostapd_data *hapd, struct sta_info *sta)
 #endif /* CONFIG_IEEE80211W */
 
 
+
+//AP-STA-DISCONNECTED 78:54:2e:05:ae:98
+//AP-STA-CONNECTED 78:54:2e:05:ae:98
+void wifi_connected_hook(char *msg, char *buf)
+{
+	char cmd[128];
+	sprintf(cmd, "/tmp/wifi_connected_hook.sh %s %s", msg, buf);
+	syslog(LOG_INFO, cmd);
+	system(cmd);
+}
+
 void ap_sta_set_authorized(struct hostapd_data *hapd, struct sta_info *sta,
 			   int authorized)
 {
@@ -1217,6 +1229,8 @@ void ap_sta_set_authorized(struct hostapd_data *hapd, struct sta_info *sta,
 		wpa_msg(hapd->msg_ctx, MSG_INFO, AP_STA_CONNECTED "%s%s",
 			buf, ip_addr);
 
+		wifi_connected_hook(AP_STA_CONNECTED, buf);
+
 		if (hapd->msg_ctx_parent &&
 		    hapd->msg_ctx_parent != hapd->msg_ctx)
 			wpa_msg_no_global(hapd->msg_ctx_parent, MSG_INFO,
@@ -1225,6 +1239,8 @@ void ap_sta_set_authorized(struct hostapd_data *hapd, struct sta_info *sta,
 	} else {
 		wpa_msg(hapd->msg_ctx, MSG_INFO, AP_STA_DISCONNECTED "%s", buf);
 
+		wifi_connected_hook(AP_STA_DISCONNECTED, buf);
+
 		if (hapd->msg_ctx_parent &&
 		    hapd->msg_ctx_parent != hapd->msg_ctx)
 			wpa_msg_no_global(hapd->msg_ctx_parent, MSG_INFO,
